fedimint-api/src/macros.rs

```diff
@@ -0,0 +1,63 @@
+/// Define "dyn newtype" (a newtype over `dyn Trait`)
+///
+/// This is a simple pattern that make working with `dyn Trait`s
+/// easier, by hiding their details.
+///
+/// A "dyn newtype" `Deref`s to the underlying`&dyn Trait`, making
+/// it easy to access the encapsulated operations, while hiding
+/// the boxing details.
+#[macro_export]
+macro_rules! dyn_newtype_define {
+    (   $(#[$outer:meta])*
+        $name:ident($container:ident<$trait:ident>)
+    ) => {
+        $(#[$outer])*
+        pub struct $name($container<dyn $trait + Send + Sync + 'static>);
+
+        impl std::ops::Deref for $name {
+            type Target = dyn $trait + Send + Sync + 'static;
+
+            fn deref(&self) -> &<Self as std::ops::Deref>::Target {
+                &*self.0
+            }
+
+        }
+
+        impl<I> From<I> for $name
+        where
+            I: $trait + Send + Sync + 'static,
+        {
+            fn from(i: I) -> Self {
+                Self($container::new(i))
```

Comment:
Is this wrapping the database in an `Arc` so it can be cloned?  What happens for a `MemDatabase`?

Reply:
Yes. My understanding is that database's API needs to accommodate concurrent use.

Reply:
> What happens for a `MemDatabase`?

It already is an `Arc<Mutex<BTreeMap<Vec<u8>, Vec<u8>>>>` internally. So maybe we can get rid of the `Arc` internally?